File: LeetCode/019_RemoveNthNodeFromEnd/removeNthNodeFromEnd.py


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head, n):
        slow = fast = head

        for _ in range(n):
            fast = fast.next

        if not fast: return head.next


        while fast.next:
            fast = fast.next
            slow = slow.next

        slow.next = slow.next.next


        return head
    

node5 = ListNode(5)
node4 = ListNode(4, node5)
node3 = ListNode(3, node4)
node2 = ListNode(2, node3)
head = ListNode(1, node2)

            
sol = Solution()

new_head = sol.removeNthFromEnd(head, n = 2)

current = new_head
while current:
    print(current.val, end=" -> ")
    current = current.next

File: LeetCode/019_RemoveNthNodeFromEnd/removeNthNodeFromEnd.ts


// Given the head of a linked list, remove the nth node from the end of the list and return its head.

//  Definition for singly-linked list.
 class ListNode {
     val: number
     next: ListNode | null
     constructor(val?: number, next?: ListNode | null) {
         this.val = (val===undefined ? 0 : val)
         this.next = (next===undefined ? null : next)
     }
 }
 

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    let dummy:ListNode = new ListNode(0);
    dummy.next = head;
    let count:number = 0;
    let current: ListNode | null = head;

    while (current !== null){
        count++;
        current = current.next;
    }

    count -= n;
    current = dummy;

    while (count > 0){
        count--;
        current = current.next!;
    }

    current.next = current.next!.next;

    return head;
};